### DISCUSSION TITLE
Some molecules returning 'zeroed' pdb coordinates during ligand preparation

### DISCUSSION THREAD

**Discussion by lynden95** [rdkit/rdkit#6149] [Q&A] [answered]
Hey! First time posting here, so hello everyone :)

I am trying to run a virtual screening cascade, starting with a .csv file containing standardised smiles strings and a Chembl ID. During the ligand prep step, I am creating molfiles from the smiles strings, adding hydrogens and embedding them as 3D coordinates using RDKit. I am then using openbabel to produce .pdb files, protonate, then convert the .pdb files to .sdf files. The individual .sdf files are then merged into one large .sdf file which can be used for docking of a large number of compounds.

However, some of the smiles strings are returning completely zeroed coordinates when I produce the pdb file, which I believe is occurring at the 'embed molecule' stage. These '0' compounds are incorporated into the final .sdf file which is causing my docking software to crash! 

I was wondering if there is a way to identify when this happens and remove the .pdb file so the '0' compounds are not incorporated into the .sdf file, and the docking should run smoothly? Below is the code I am currently using for this. It is worth noting that beforehand I was using default EmbedMolecule parameters (random coords = False)...

`# The below takes a .csv file with smiles strings and converts them to a single .sdf file containing all smiles strings in protonated 3D format.

def prepare_ligands_sdf(std_csv_file, pH, header_len=1, output_dir = sdf_folder, delim=',') -> list:
  
    out_sdfs = [] # Creates an empty list to fill.
    
    print(std_csv_file)
    
    with open(std_csv_file, 'r') as csv: 
        
        
        for entry in csv.readlines()[header_len:]:
            
            ID, ST_SMILES = entry.split(delim)[:2]            
           
            # Convert smiles str to 3D coordinates
            mol = Chem.MolFromSmiles(ST_SMILES)
            mol = Chem.AddHs(mol)
            ps = AllChem.ETKDGv2()
            ps.useRandomCoords = True
            AllChem.EmbedMolecule(mol, ps)
         
            
            # Ouput coords to pdb
            pdb_name = f"{output_dir}/{ID}.pdb"
            MolToPDBFile(mol, pdb_name)
           

            # Protonate according to pH, convert to .sdf
            sdf_name = f"{output_dir}/{ID}.sdf"
            ! obabel {pdb_name} -pH {pH} -O {sdf_name}
          
            os.remove(pdb_name) 
            
            out_sdfs.append(sdf_name)
            
    return out_sdfs

def merge_sdfs(out_sdfs, merged_sdf):

    mols = []
    for s in out_sdfs:
        suppl = Chem.SDMolSupplier(s)
        for  i, mol in enumerate(suppl):
            if mol is None:
                print(i)
                continue
            else:
                mols += [mol]
       
        os.remove(s)

    with Chem.SDWriter(merged_sdf) as w:
        for mol in mols:
            w.write(mol)
    

def prepare_and_merge_ligands(std_csv_file,  pH, header_len=1, output_dir='', delim=','):
    ligands = prepare_ligands_sdf(std_csv_file,  pH, header_len, output_dir, delim)
    merge_sdfs(ligands, output_dir+"/smallmols_filtered_merged.sdf")
`

**Comment by lynden95** [comment, marked as answer]
For those interested - this was resolved with the following:
`mol = Chem.MolFromSmiles(ST_SMILES)
            mol = Chem.AddHs(mol)
            ps = AllChem.ETKDGv2()
            ps.useRandomCoords = True
            result = AllChem.EmbedMolecule(mol, ps)

            if type(result) is int:
                if result == -1:
                    continue
            
            # Ouput coords to pdb
            pdb_name = f"{output_dir}/{ID}.pdb"
            MolToPDBFile(mol, pdb_name)`

**Reply by greglandrum** [reply]
As a general suggestion; you don't need to check the type of the return result from `EmbedMolecule()`, that will always be an integer, so you can just do:
```
mol = Chem.MolFromSmiles(ST_SMILES)
mol = Chem.AddHs(mol)
ps = AllChem.ETKDGv2()
ps.useRandomCoords = True
result = AllChem.EmbedMolecule(mol, ps)
if result == -1:
    continue
        
# Ouput coords to pdb
pdb_name = f"{output_dir}/{ID}.pdb"
MolToPDBFile(mol, pdb_name)`
```
